### PR TITLE
Fuyu Finetuning Example

### PR DESCRIPTION
**Pull request by ncoop57**
# What does this PR do?

This PR is about contributing an example for finetuning the recent Adept Fuyu model architecture. It adds the ability to calculate the loss for the model as well as add a new folder for image-text examples. It is currently in DRAFT model as it still has a bug due to the data collator padding and an easy way of performing evaluation. I'm opening the PR in this state so that I can get help.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

@molbap would love some help since you initially created the model!

Models:

- text models: @ArthurZucker and @younesbelkada
- vision models: @amyeroberts

Library:

- tokenizers: @ArthurZucker
- trainer: @muellerzr and @pacman100

### CODE REVIEW

**Comment by ncoop57** [comment]
The issue I am getting right now is that I am unable to pad the image patches correct it seems for batch training. The image processor doesn't seem to be doing it correct.

**Comment by ncoop57** [comment]
I thought it would be as simple as modifying variable_size var to false: https://github.com/huggingface/transformers/blob/main/src/transformers/models/fuyu/processing_fuyu.py#L508C1-L508C37

And that seems to patch the image patches correct, but the `image_patches_indices` var doesn't seem to be updated properly

**Comment by Junemengyuan** [comment]
> The issue I am getting right now is that I am unable to pad the image patches correct it seems for batch training. The image processor doesn't seem to be doing it correct.

Yeah, the current image processor seems only outputs batch-size=1 result

**Comment by ArthurZucker** [comment]
Hey! Thanks a lot for wanting to contribute! We usually share links to training scripts like in the ressource section (like [ this one](https://twitter.com/m_olbap/status/1715362812757999991)), but we don't add them to the repo for every new model! @molbap is working on a fix for the image processor see #27007

**Comment by ncoop57** [comment]
Heyya @ArthurZucker ! In that case, should I close this PR or should this PR be the first of multi-modal model training examples? Seems like the space is getting super interesting and so it might be nice for people to see example scripts in the official repo for huggingface

**Comment by gagan3012** [comment]
Hi Is this working now?

**Comment by Carolinabanana** [comment]
This is working for me, but only with batch size 1. Using 4-bit Lora training this does fit on a 24GB GPU.

Using higher than batch 1 gives the error I reported in: #27255

**Comment by molbap** [comment]
Hi @ncoop57! It was mentioned before not sure if this PR should be merged here but it's a great resource. We don't have official training scripts, but we updated the `FuyuProcessor` has been updated with batching + attention masks with left-padding, see https://github.com/huggingface/transformers/blob/88832c01c8a962b653874c4ce4ed8df5783ac5cd/src/transformers/models/fuyu/processing_fuyu.py#L335C1-L383C1. Is that factored in for your loss calculation? I'm not 100% sure about the code but I think it would need to be adjusted a bit, such as 

```python
def collate_fn(examples):
    texts = [e[text_column_name] for e in examples]
    images = [e[image_column_name] for e in examples]
    output = processor(
        text=texts,
        images=images,
        padding="max_length",
        truncation=True
    )
    first_non_padded = (output["input_ids"] != tokenizer.pad_token_id).nonzero(as_tuple=False)
    first_non_padded_indices = first_non_padded[:, 1].view(-1, 1)
    labels = torch.full_like(output["input_ids"], -100)
    for i, index in enumerate(first_non_padded_indices):
        labels[i, index:] = output["input_ids"][i, index:]
    output["labels"] = labels
    return output

```
let me know what you think!

**Comment by ncoop57** [comment]
Super happy to see more excitement around this example script! Thanks for the loss pointer, I'm looking into it. I had the same error as @Carolinabanana mentioned but only when using the Trainer object, not when doing it myself. I did try to account for padding/image patches because I don't think we want to train on the prediction of the image patch indices, but I did forget to take into account the left padding!

**Comment by infamix** [comment]
The script doesn't support VQA, does it?

**Comment by cysmnl** [comment]
Can it fit in a 80GB card with bfloat16 for tuning? I faced a CUDA memory issue when running `run_fuyu_no_trainer.py` on a single A100 card.

**Comment by zhaoyukoon** [comment]
Can anyone provide a train script? Thanks!

**Comment by amyeroberts** [comment]
Hi @ncoop57 Any update on the script and readiness for PR review? Already looks very good! Let us know if you need any help adding it to the library

**Comment by ncoop57** [comment]
AH sadly haven't had a chance to look back into this. Luckily Ill be off all next week so gonna add this to my list to complete and will reach out with help that I need, ty!

**Comment by ncoop57** [comment]
@amyeroberts think I got it fully in and ready for review. @molbap if you have some time, I'd love for you to checkout how I am handling the left padding:

```python
def collate_fn(examples):
        texts = [e[text_column_name] for e in examples]
        images = [e[image_column_name] for e in examples]
        output = processor(
            text=texts, images=images, padding="max_length", truncation=True
        )
        position = (output["input_ids"] == tokenizer.vocab["<s>"]).nonzero(
            as_tuple=True
        )[0][
            0
        ]  # This gets the index of the first '1' in the tensor by passing the left padding
        output["labels"] = torch.full_like(
            output["input_ids"], -100
        )  # This creates a tensor filled with -100
        output["labels"][position:] = output["input_ids"][position:]
        return output
```

Let me know what else is needed to get this merged in 🤓

**Comment by ncoop57** [comment]
I wasn't able to get the Trainer working so right now this is just with a standard training loop and uses accelerate to handle all the FSDP stuff

**Comment by ncoop57** [comment]
@amyeroberts looks like I got all the CI stuff working 🤓 lemme know if any other stuff are needed

**Comment by ncoop57** [comment]
maybe one tweak would be renaming the folder to "visual-language-modeling"?

File: examples/pytorch/image-text/run_fuyu_no_trainer.py [hunk 1]
```diff
@@ -0,0 +1,778 @@
...
+    FuyuProcessor,
+    SchedulerType,
+    get_scheduler,
+)
+from transformers.utils import check_min_version, send_example_telemetry
+from transformers.utils.versions import require_version
+
+
+# Will error if the minimal version of Transformers is not installed. Remove at your own risks.
+check_min_version("4.36.0.dev0")
```
Comment (amyeroberts) [inline examples/pytorch/image-text/run_fuyu_no_trainer.py:60]:
```suggestion
check_min_version("4.38.0.dev0")
```

File: examples/pytorch/image-text/run_fuyu_no_trainer.py [hunk 2]
```diff
@@ -0,0 +1,778 @@
...
+                args.dataset_config_name,
+                split=f"test[{args.validation_split_percentage}%:]",
+            )
+            raw_datasets["validation"] = load_dataset(
+                args.dataset_name,
+                args.dataset_config_name,
+                split=f"test[:{args.validation_split_percentage}%]",
+            )
+        elif "validation" not in raw_datasets.keys():
+            raw_datasets["validation"] = load_dataset(
```
Comment (amyeroberts) [inline examples/pytorch/image-text/run_fuyu_no_trainer.py:393]: Two comments: 
* Why call `load_dataset` twice if `args.dataset_name == "facebook/winoground`? 
* Why this specific winoground logic? The scripts should be as simple as possible and not handle all different use cases.

File: examples/pytorch/image-text/run_fuyu_no_trainer.py [hunk 3]
```diff
@@ -0,0 +1,778 @@
...
+    elif args.model_name_or_path:
+        tokenizer = AutoTokenizer.from_pretrained(
+            args.model_name_or_path,
+            use_fast=not args.use_slow_tokenizer,
+            trust_remote_code=args.trust_remote_code,
+        )
+    else:
+        raise ValueError(
+            "You are instantiating a new tokenizer from scratch. This is not supported by this script. "
+            "You can do it from another script, save it, and load it from here, using --tokenizer_name."
```
Comment (amyeroberts) [inline examples/pytorch/image-text/run_fuyu_no_trainer.py:466]: We don't need to load a tokenizer if we're loading a processor. The processor loading should be loaded using this if/elif/else structure here instead and using `AutoProcessor`

File: examples/pytorch/image-text/run_fuyu_no_trainer.py [hunk 4]
```diff
@@ -0,0 +1,778 @@
...
+    # Preprocessing the datasets.
+    text_column_name = args.dataset_caption_column_name
+    image_column_name = args.dataset_image_column_name
+
+    def filter_corrupt_images(examples):
+        """remove problematic images"""
+        valid_images = []
+        for i in range(0, len(examples[image_column_name])):
+            try:
+                processor(text="test", images=[examples[image_column_name][i]])
```
Comment (amyeroberts) [inline examples/pytorch/image-text/run_fuyu_no_trainer.py:500]: Why pass them through the processor here? Trying to open them with `Image.open` should be enough to check. A failure here doesn't mean the image itself is problematic.

Comment (amyeroberts) [inline examples/pytorch/image-text/run_fuyu_no_trainer.py:500]: * range automatically starts on the 0th index
* You can iterate directly on the dataset

```suggestion
        for image in examples[image_column_name]:
            try:
                Image.open(image)
```

File: examples/pytorch/image-text/run_fuyu_no_trainer.py [hunk 5]
```diff
@@ -0,0 +1,778 @@
...
+                valid_images.append(False)
+        return valid_images
+
+    with accelerator.main_process_first():
+        raw_datasets = raw_datasets.filter(
+            filter_corrupt_images,
+            batched=True,
+        )
+        # resize images
+        resize = Resize((224, 224))
```
Comment (amyeroberts) [inline examples/pytorch/image-text/run_fuyu_no_trainer.py:512]: The size here should be taken from the image processor config: 

```py
size_dict = processor.image_processor.size
size = (size_dict["height"], size_dict["width"])
resize = Resize(size)
```

Comment (amyeroberts) [inline examples/pytorch/image-text/run_fuyu_no_trainer.py:512]: Why is this the only transformation applied? I'd expect some augmentation in a training script. You can look at the [image classification script as an example](https://github.com/huggingface/transformers/blob/39c3c0a72af6fbda5614dde02ff236069bb79827/examples/pytorch/image-classification/run_image_classification_no_trainer.py#L364).

File: examples/pytorch/image-text/run_fuyu_no_trainer.py [hunk 6]
```diff
@@ -0,0 +1,778 @@
...
+        raw_datasets = raw_datasets.map(
+            lambda x: {"image_0": resize(x["image_0"])},
+            batched=False,
+        )
+
+    if args.block_size is None:
+        block_size = tokenizer.model_max_length
```
Comment (amyeroberts) [inline examples/pytorch/image-text/run_fuyu_no_trainer.py:519]: This should use the processor's tokenizer

**Review by amyeroberts — COMMENTED** [review]
Thanks for all the work adding this! 

Overall looking good. Made some general comments on the structure. There's some places which have been copied across from other scripts but need adapting to this one. 

Before we can publish this script, we'll need a demonstration of a model which has been trained for a few epochs and pushed to the hub successfully. If/when you have thant, could you share in the PR description?

File: examples/pytorch/image-text/run_fuyu_no_trainer.py (hunk 5 above)
Comment (ncoop57) [reply examples/pytorch/image-text/run_fuyu_no_trainer.py:512]: Actually figured out the processor already handles resizing, so need to do something like:
```
processor = FuyuProcessor.from_pretrained(args.model_name_or_path, size = {"height": 480, "width": 480})
```

**Comment by ncoop57** [comment]
Awesome, thanks for the review and recommendations! 🤓 I've been working on finetuning using the https://huggingface.co/datasets/HuggingFaceM4/WebSight dataset, so once I have the model trained I'll make sure to add it here and to the README!

File: examples/pytorch/image-text/run_fuyu_no_trainer.py (hunk 5 above)
Comment (amyeroberts) [reply examples/pytorch/image-text/run_fuyu_no_trainer.py:512]: Yes, but it doesn't do augmentation. I'd recommend not using the processor to process the images.
